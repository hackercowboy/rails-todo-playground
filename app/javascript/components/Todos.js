import React, { useState, useEffect } from 'react';
import axios from 'axios';
import ErrorList from './ErrorList';
import TodoForm from './TodoForm';
import TodoList from './TodoList';

const e = React.createElement;

function headers() {
  return {
    'X-CSRF-Token': document.head.querySelector('[name=csrf-token]').content,
    'Content-Type': 'application/json',
  };
}

function Todos() {
  const [todos, setTodos] = useState([]);
  const [errors, setErrors] = useState();

  const loadData = async () => {
    const response = await axios.get('/api/v1/todos');
    setTodos(response.data.todos);
  };

  useEffect(() => {
    loadData();
  }, []);

  const handleOnSubmit = async (todo) => {
    setErrors();
    await axios.post('/api/v1/todos', todo, { headers: headers() });
    await loadData();
  };

  const handleOnError = async (nextErrors) => {
    setErrors(nextErrors);
  };

  const handleOnDone = async (id) => {
    await axios.patch(`/api/v1/todos/${id}`, { status: 'done' }, { headers: headers() });
    await loadData();
  };

  const handleOnDelete = async (id) => {
    await axios.delete(`/api/v1/todos/${id}`, { headers: headers() });
    await loadData();
  };

  return e(
    React.Fragment,
    null,
    [e(TodoForm, { key: 'form', onSubmit: handleOnSubmit, onError: handleOnError }),
      e(ErrorList, { key: 'errors', errors }),
      e(TodoList, {
        key: 'list', todos, onDelete: handleOnDelete, onDone: handleOnDone,
      })],
  );
}

export default Todos;
