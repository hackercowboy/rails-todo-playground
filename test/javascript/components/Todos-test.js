import React from 'react';
import { shallow } from 'enzyme';
import axios from 'axios';

import Todos from '../../../app/javascript/components/Todos';

jest.mock('axios');

describe('<Todos/>', () => {
  it('should render', async () => {
    axios.get.mockResolvedValue({ data: { todos: [{ id: 21, text: 'Test' }] } });
    const wrapper = await shallow(<Todos/>);
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should handleOnSubmit', async () => {
    axios.get.mockResolvedValue({ data: { todos: [] } });
    document.head.querySelector = () => ({ content: 'csrf-token-abcd' });

    const wrapper = await shallow(<Todos/>);
    wrapper.find('TodoForm').props().onSubmit({ text: 'Test' });
    expect(axios.post).toBeCalledWith('/api/v1/todos', { text: 'Test' }, { headers: { 'Content-Type': 'application/json', 'X-CSRF-Token': 'csrf-token-abcd' } });
  });

  it('should handleErrors', async () => {
    axios.get.mockResolvedValue({ data: { todos: [] } });
    const wrapper = await shallow(<Todos/>);
    wrapper.find('TodoForm').props().onError(['Error 1', 'Error 2']);
  });

  it('should handleOnDone', async () => {
    axios.get.mockResolvedValue({ data: { todos: [] } });
    document.head.querySelector = () => ({ content: 'csrf-token-abcd' });

    const wrapper = await shallow(<Todos/>);
    wrapper.find('TodoList').props().onDone(43);
    expect(axios.patch).toBeCalledWith('/api/v1/todos/43', { status: 'done' }, { headers: { 'Content-Type': 'application/json', 'X-CSRF-Token': 'csrf-token-abcd' } });
  });

  it('should handleOnDelete', async () => {
    axios.get.mockResolvedValue({ data: { todos: [] } });
    document.head.querySelector = () => ({ content: 'csrf-token-abcd' });

    const wrapper = await shallow(<Todos/>);
    wrapper.find('TodoList').props().onDelete(43);
    expect(axios.delete).toBeCalledWith('/api/v1/todos/43', { headers: { 'Content-Type': 'application/json', 'X-CSRF-Token': 'csrf-token-abcd' } });
  });
});
