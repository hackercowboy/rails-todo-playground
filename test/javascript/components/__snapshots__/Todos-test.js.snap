// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todos/> should render 1`] = `
"<Fragment>
  <TodoForm onSubmit={[Function: handleOnSubmit]} onError={[Function: handleOnError]} />
  <ErrorList errors={[undefined]} />
  <TodoList todos={{...}} onDelete={[Function: handleOnDelete]} onDone={[Function: handleOnDone]} />
</Fragment>"
`;
